ocaml/sdk-gen/go/gen_go_helper.ml

```diff
@@ -169,32 +279,598 @@ module Json = struct
+            [
+              ("type", `String t')
+            ; ("func_partial_type", `String (get_fp_type t))
+            ]
+
+  let group_params params =
+    List.fold_left
+      (fun groups (param_release, info) ->
+        match groups with
+        | [] ->
+            [(param_release, [info])]
+        | group :: tl -> (
+          match group with
+          | pr, infos when pr = param_release ->
+              (pr, info :: infos) :: tl
+          | _ ->
+              (param_release, [info]) :: groups
+        )
+      )
+      [] params
+
+  let scan_left func init list =
+    let rec aux func init list result =
+      match list with
+      | [] ->
+          result
+      | x :: xs ->
+          let init = func init x in
+          aux func init xs (init :: result)
+    in
+    List.rev @@ aux func init list []
+
+  let combine_groups groups =
+    let scans =
+      scan_left
+        (fun (len, group) (_, infos) -> (len + List.length infos, group @ infos))
+        (0, []) groups
+    in
+    match scans with [] -> [] | [x] -> [x] | _ :: xs -> xs
+
+  let rec get_last = function
+    | [] ->
+        failwith "empty list"
+    | [x] ->
+        x
+    | _ :: tl ->
+        get_last tl
+
+  let of_param_groups class_name params =
+    let name_internal name =
+      let name = name |> snake_to_camel |> String.uncapitalize_ascii in
+      match name with "type" -> "typeKey" | "interface" -> "inter" | _ -> name
+    in
+    let base_assoc_list (t, name, doc, type_name) =
+      [
+        ("session", `Bool (name = "session_id"))
+      ; ("type", `String t)
+      ; ("name", `String name)
+      ; ("name_internal", `String (name_internal name))
+      ; ("doc", `String doc)
+      ; ("func_partial_type", `String type_name)
+      ]
+    in
+    let deal_with_logout name =
+      match (class_name, name) with
+      | "session", "session_id" ->
+          ([("param_ignore", `Bool true); ("session_class", `Bool true)], true)
+      | "session", _ ->
+          ([("param_ignore", `Bool false); ("session_class", `Bool true)], false)
+      | _ ->
+          ([("param_ignore", `Null); ("session_class", `Null)], false)
+    in
+    let get_assoc_list (t, name, doc, type_name) =
+      let fields, could_ignore = deal_with_logout name in
+      (fields @ base_assoc_list (t, name, doc, type_name), could_ignore)
+    in
+    let rec add_first = function
+      | head :: rest ->
+          let assoc_list, could_ignore = get_assoc_list head in
+          let assoc_list = ("first", `Bool (not could_ignore)) :: assoc_list in
+          let others =
+            if could_ignore then
+              add_first rest
+            else
+              List.map
+                (fun item ->
+                  let assoc_list, _ = get_assoc_list item in
+                  `O (("first", `Bool false) :: assoc_list)
+                )
+                rest
+          in
+          `O assoc_list :: others
+      | [] ->
+          []
+    in
+    let groups =
```

Comment:
Would it be easier to do the following:
```
[p0; p1; p2; p3; p4] -> [[p0; p1]; [p0; p1; p2]; [p0; p1; p2; p3; p4]]
```
This could be done by:
```
List.map (fun release List.filter (p -> p.release <= release) params) release_ordered_list
```
After this:
```
l > List.concat |> List.map of_message ...
```

Reply:
These related code has been split to the following PR which will raised after this one got merged.